### PR TITLE
Redesign add plugins page

### PR DESCRIPTION
**Pull request by thornbill**
**Changes**
Implements a completely overhauled plugin installation page.

Screenshot 2024-07-11 at 16-27-50 Jellyfin

Screenshot 2024-07-11 at 16-29-44 Jellyfin

Additional features that could be added in the future:

* Restart button when the plugin status is restart needed
* Indicator when a plugin update is available

**Issues**
N/A

### CODE REVIEW

**Comment by thornbill** [comment]
Not sure why the preview link comment is not being posted, but you can find it here for testing:

Action run: https://github.com/jellyfin/jellyfin-web/actions/runs/9911940022
CF Pages url: https://ad315c5f.jellyfin-web.pages.dev/

**Comment by nielsvanvelzen** [comment]
Did some testing on the CF deployment. The design looks great and it's a real improvement from the current page, Here are my findings. Some are nitpicks, other are most likely due to how the API works.

- I added my own server so I could sign in as an admin user. After refreshing the page it went back to the default (unstable demo). Unrelated to this issue but probably needs some attention.
- A plugin with extra pages (Like the playback reporting plugin) should probably have those pages added as buttons, just like the "settings" button.
- Pressing the "install" button doesn't give any feedback until the install is completed. Considering this task may take some time it would be nice to show a progress indicator or something like that
- Pressing the "install" button should _probably_ show a confirmation dialog first (like uninstall)
- When I install a plugin and then uninstall it again it still shows the "The plugin has been successfully installed." message
- The "The plugin has been successfully installed." message could probably have a "restart" button on the right side
- The install buttons in the revision history should be renamed to "downgrade" when there is a newer version installed, with a dialog to confirm
- When the plugin is auto-updated but still needs a restart, the status is "restart". Could use a restart button to make it easier (maybe add an message just like when the plugin is installed).
- We should probably open this page when clicking on a plugin from the "My Plugins" page

**Comment by thornbill** [comment]
- I added my own server so I could sign in as an admin user. After refreshing the page it went back to the default (unstable demo). Unrelated to this issue but probably needs some attention.
  - Likely part of larger issue with our multi-server handling.
- A plugin with extra pages (Like the playback reporting plugin) should probably have those pages added as buttons, just like the "settings" button.
  - Probably not currently possible. Plugins seem to also use configuration pages to expose static assets (js files) and don't assign the different pages unique display names.
- Pressing the "install" button doesn't give any feedback until the install is completed. Considering this task may take some time it would be nice to show a progress indicator or something like that
  - I was planning to use the MUI [LoadingButton](https://mui.com/material-ui/react-button/#loading-button) for this but it requires installing their "labs" package so I was going to do it in a separate PR.
- Pressing the "install" button should _probably_ show a confirmation dialog first (like uninstall)
  - This keeps the current behavior. A confirmation/warning prompt is displayed if installing from a third-party repository.
- When I install a plugin and then uninstall it again it still shows the "The plugin has been successfully installed." message
  - They use different APIs so it doesn't reset automatically, but should be an easy fix.
- The "The plugin has been successfully installed." message could probably have a "restart" button on the right side
  - Planned for a future PR
- The install buttons in the revision history should be renamed to "downgrade" when there is a newer version installed, with a dialog to confirm
  - Seems like a good enhancement for a future PR... We would need to pull in a dependency or write a utility to compare versions.
- When the plugin is auto-updated but still needs a restart, the status is "restart". Could use a restart button to make it easier (maybe add an message just like when the plugin is installed).
  - Planned for a future PR
- We should probably open this page when clicking on a plugin from the "My Plugins" page
  - Planned for a future PR (the page doesn't handle built-in plugins great atm)

File: src/apps/dashboard/routes/plugins/plugin.tsx [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+    }, [ disablePlugin, enablePlugin, installPlugin, pluginDetails, uninstallPlugin ]);
+
+    /** Close the uninstall confirmation dialog */
+    const onCloseUninstallConfirmDialog = useCallback(() => {
+        setIsUninstallConfirmOpen(false);
+    }, []);
+
+    useEffect(() => {
+        setIsLoading(isConfigurationPagesLoading || isPackageInfoLoading || isPluginsLoading);
+    }, [ isConfigurationPagesLoading, isPackageInfoLoading, isPluginsLoading ]);
```
Comment (grafixeyehero) [inline src/apps/dashboard/routes/plugins/plugin.tsx:250]:
```suggestion
    const isLoading =
        isConfigurationPagesLoading || isPackageInfoLoading || isPluginsLoading;
```

File: src/apps/dashboard/routes/plugins/plugin.tsx [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+
+    useEffect(() => {
+        setPluginName(searchParams.get('name') ?? undefined);
+    }, [ searchParams ]);
```
Comment (grafixeyehero) [inline src/apps/dashboard/routes/plugins/plugin.tsx:254]:
```suggestion
    const pluginName = searchParams.get('name') ?? undefined;
```

File: src/apps/dashboard/routes/plugins/plugin.tsx [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+                    ?? pluginInfo?.Status !== PluginStatus.Disabled,
+                name: pluginName || pluginInfo?.Name || packageInfo?.name,
+                owner: packageInfo?.owner,
+                status: pluginInfo?.Status,
+                configurationPage: findBestConfigurationPage(configurationPages || [], pluginId),
+                version,
+                versions: packageInfo?.versions || []
+            });
+        }
+    }, [ api, configurationPages, isEnabledOverride, isPluginsLoading, pluginId, packageInfo, pluginName, plugins ]);
```
Comment (grafixeyehero) [inline src/apps/dashboard/routes/plugins/plugin.tsx:296]:
```suggestion
    const pluginDetails = useMemo(() => {
        if (pluginId && !isPluginsLoading) {
            const pluginInfo = findBestPluginInfo(pluginId, plugins);

            let version;
            if (pluginInfo) {
                // Find the installed version
                const repoVersion = packageInfo?.versions?.find(v => v.version === pluginInfo.Version);
                version = repoVersion || {
                    version: pluginInfo.Version,
                    VersionNumber: pluginInfo.Version
                };
            } else {
                // Use the latest version
                version = packageInfo?.versions?.[0];
            }

            let imageUrl;
            if (pluginInfo?.HasImage) {
                imageUrl = api?.axiosInstance.getUri({
                    baseURL: api.basePath,
                    url: `/Plugins/${pluginInfo.Id}/${pluginInfo.Version}/Image`
                });
            }

            return {
                canUninstall: !!pluginInfo?.CanUninstall,
                description: pluginInfo?.Description || packageInfo?.description || packageInfo?.overview,
                id: pluginId,
                imageUrl: imageUrl || packageInfo?.imageUrl || undefined,
                isEnabled: (isEnabledOverride && pluginInfo?.Status === PluginStatus.Restart)
                    ?? pluginInfo?.Status !== PluginStatus.Disabled,
                name: pluginName || pluginInfo?.Name || packageInfo?.name,
                owner: packageInfo?.owner,
                status: pluginInfo?.Status,
                configurationPage: findBestConfigurationPage(configurationPages || [], pluginId),
                version,
                versions: packageInfo?.versions || []
            };
        }
    }, [api?.axiosInstance, api?.basePath, configurationPages, isEnabledOverride, isPluginsLoading, packageInfo?.description, packageInfo?.imageUrl, packageInfo?.name, packageInfo?.overview, packageInfo?.owner, packageInfo?.versions, pluginId, pluginName, plugins]);
```

**Review by grafixeyehero — CHANGES_REQUESTED** [review]

File: src/apps/dashboard/routes/plugins/plugin.tsx (hunk 3 above)
Comment (grafixeyehero) [reply src/apps/dashboard/routes/plugins/plugin.tsx:296]: move the pluginDetails change to line 85
this change reduces unnecessary re-renders.

Before:
Screenshot 2024-07-14 214731

After:
Screenshot 2024-07-14 213816

**Comment by grafixeyehero** [comment]
left those change to the top.
Those change reduces unnecessary re-renders
Before:
Screenshot 2024-07-14 214731

After:
Screenshot 2024-07-14 213816

**Review by grafixeyehero — CHANGES_REQUESTED** [review]

**Review by grafixeyehero — APPROVED** [review]

**Comment by thornbill** [comment]
The version list expands when you click on a version to show the change log and an install button.

**Pull request merged**